### PR TITLE
Document solana_program::instruction

### PR DESCRIPTION
**Pull request by brson**
#### Problem

Hackers love docs and they don't get enough of them.

#### Summary of Changes

This adds docs and examples for everything in the instruction module. Please review the "Specifying account metadata" section closely.

I made the `checked_add` function hidden as it seems to be intended for internal use.

cc @askibin I appreciated your review of my last doc patch.

### CODE REVIEW

File: sdk/program/src/instruction.rs [hunk 1]
```diff
@@ -240,80 +252,312 @@ pub enum InstructionError {
     // conversions must also be added
 }
 
+/// A directive for a single invocation of a Solana program.
+///
+/// An instruction specifies which program it is calling, which accounts it may
+/// read or modify, and additional data that serves as input to the program. One
+/// ore more instructions are included in transactions submitted by Solana
```
Comment (CriesofCarrots) [inline sdk/program/src/instruction.rs:259]:
```suggestion
/// or more instructions are included in transactions submitted by Solana
```

File: sdk/program/src/instruction.rs [hunk 2]
```diff
@@ -240,80 +252,312 @@ pub enum InstructionError {
...
+///
+/// When building a transaction, the Solana runtime coalesces all accounts used
+/// by all instructions in that transaction, along with accounts and permissions
+/// required by the runtime, into a single account list. Some accounts and
+/// account permissions required by the runtime to process a transaction are
+/// _not_ required to be included in an `Instruction`s account list. These
+/// include:
+///
+/// - The program ID &mdash; it is a separate field of `Instruction`
+/// - The transaction's fee-paing acount &mdash; it is added during [`Message`]
```
Comment (CriesofCarrots) [inline sdk/program/src/instruction.rs:305]:
```suggestion
/// - The transaction's fee-paying account &mdash; it is added during [`Message`]
```

File: sdk/program/src/instruction.rs [hunk 3]
```diff
@@ -240,80 +252,312 @@ pub enum InstructionError {
...
 #[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
 pub struct AccountMeta {
-    /// An account's public key
+    /// An account's public key.
     pub pubkey: Pubkey,
-    /// True if an Instruction requires a Transaction signature matching `pubkey`.
+    /// True if an `Instruction` requires a `Transaction` signature matching `pubkey`.
     pub is_signer: bool,
-    /// True if the `pubkey` can be loaded as a read-write account.
+    /// True if the account may be mutated, either its data or metadata, during program execution.
```
Comment (CriesofCarrots) [inline sdk/program/src/instruction.rs:529]:
```suggestion
    /// True if the account data or metadata may be mutated during program execution.
```

**Review by CriesofCarrots — COMMENTED** [review]
Couple nits, otherwise looks great

**Comment by brson** [comment]
Suggestions committed. Thanks for catching my typos. Great eyes.

**Review by CriesofCarrots — APPROVED** [review]
Excellent, thank you!

**Pull request merged**